### PR TITLE
Block bindings: Improve performance by memo postypes slug.

### PR DESCRIPTION
**Pull request by cbravobernal**
## What?

Improve the perfomance of the postTypesSlugs by memoizing it.

trunk:
trunk

PR:
Screenshot 2024-09-27 at 14 19 19

### CODE REVIEW

File: packages/editor/src/components/provider/index.js [hunk 1]
```diff
@@ -194,8 +194,11 @@ export const ExperimentalEditorProvider = withRegistryProvider(
 					postContext.postType = 'post';
 				} else if ( post.slug.split( '-' )[ 0 ] === 'single' ) {
 					// If the slug is single-{postType}, infer the post type from the slug.
-					const postTypesSlugs =
-						postTypes?.map( ( entity ) => entity.slug ) || [];
+					const postTypesSlugs = useMemo( () => {
+						return (
+							postTypes?.map( ( entity ) => entity.slug ) || []
+						);
+					}, [] );
```
Comment (Mamaduka) [inline packages/editor/src/components/provider/index.js]: Using hooks inside condition is against [React Hook rules](https://react.dev/warnings/invalid-hook-call-warning), and I don't think nesting `useMemo` calls is a good idea.

We should move `postTypesSlugs` map outside of current memo function.

**Comment by cbravobernal** [comment]
Is the PR increasing the site-editor writing a 7%?
  
Screenshot 2024-09-27 at 18 35 05

**Comment by SantosGuillamot** [comment]
> Is the PR increasing the site-editor writing a 7%?

It seems that specific metric varies a lot. Taking a look other random PRs:
* 10,77%: [link](https://github.com/WordPress/gutenberg/actions/runs/11174756896).
* 12,76%: [link](https://github.com/WordPress/gutenberg/actions/runs/11143125022).
* -4.81%: [link](https://github.com/WordPress/gutenberg/actions/runs/11164016604).

I'm more worried about `firstBlock` because it is the one the original PR was increasing: [link](https://github.com/WordPress/gutenberg/pull/65062#discussion_r1753862288). I've been taking a look at the results, and I can see a spike in that metric around the time that PR was merged:

Screenshot 2024-10-04 at 08 51 16

Taking a look at the PRs that got merged before that one, it seems it could affect that.

| PR | First block - post-editor | First block - site-editor |
|----------|----------|----------|
| [**Original PR**](https://github.com/WordPress/gutenberg/actions/runs/10716756125) | 3723.2 ms | 4230.7 ms |
| [#64819](https://github.com/WordPress/gutenberg/actions/runs/10791146949) | 3369.25 ms | 3951.95 ms |
| [#65222](https://github.com/WordPress/gutenberg/actions/runs/10804272070) | 3394.05 ms | 3913.1 ms |
| [#65117](https://github.com/WordPress/gutenberg/actions/runs/10807337481) | 3394 ms | 3788.75 ms |
| [#65231](https://github.com/WordPress/gutenberg/actions/runs/10807957473) | 3554.15 ms | 3719.2 ms |

According to this data, it is a possibility that the original PR introduced a regression on this aspect. I'd like to:
1. Verify that the original pull request actually created the problem. Maybe testing a revert?
2. If that's the case, fix or at least palliate that.

Taking a look at the performance metrics from this PR, it seems the `firstBlock` metric doesn't improve.

**Comment by SantosGuillamot** [comment]
I've opened [this other pull request](https://github.com/WordPress/gutenberg/pull/66101) to potentially palliate the `firstBlock` metric. I believe both PRs can be handled independently.

**Comment by Mamaduka** [comment]
@SantosGuillamot, can we close this PR now that #66101 is merged?

**Comment by SantosGuillamot** [comment]
> can we close this PR now that https://github.com/WordPress/gutenberg/pull/66101 is merged?

The biggest performance impact seems to be solved after that PR, yes. Having said that, if you think the logic in this pull request can still help, I'm fine with it. I don't have much experience with this topic, so I don't have a strong opinion.

**Comment by cbravobernal** [comment]
Closing in favor of #66101

**Pull request closed without merging**